### ISSUE TITLE
Global Type Declaration for file loader packages

### ISSUE THREAD

**Issue by nbouvrette** [microsoft/TypeScript#49124]
# Suggestion

When writing a file loader package, let's say to add a new file type to import (e.g., `*.txt`), it would be great that when someone installs this package that there is a way that by default `*.txt` files would have the `string` type (or whatever type the file loader package determine)

## 🔍 Search Terms

global declaration

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

This is just an idea of a new "declare global" syntax:

```ts
declare global module '*.txt' {
  const text: string;
  export default text;
}
```

## 📃 Motivating Example

The popular CSS loader package could be an example where this could be useful (I'm not affiliated with this repo, just providing an example): https://www.npmjs.com/package/css-loader

## 💻 Use Cases

Any file loader out there, if you want to add the type of the file of the package, there is no way to automate this without using `npx`  or manual installation steps which is inconvenient.

**Comment by IllusionMH** [comment]
In what file you should put this declaration and what mechanism should hit TS to load that file?
If it's not in `includes` and imports from included files then why that file should be processed?

Existing `declare module '*.txt'` should be enough if TS will process file with this declaration.

**Comment by nbouvrette** [comment]
I guess it would be niche scenarios where you could rely on Webpack configs to enable imports and could benefit from having the correct type when importing those files without necessarily needing to import anything from the package providing the file loader.

I'm not too sure how this would work exactly and how many benefits VS challenges such a feature would bring?

**Comment by nbouvrette** [comment]
@andrewbranch I'm not sure I understand? Ambient modules will only apply if you import the module - what I was proposing is some sort of inherited global ambient module that would apply without importing anything.

**Comment by andrewbranch** [comment]
That’s not true. The pattern ambient module will apply as soon as the file that declares it enters your program for any reason. That can be due to an import, a triple-slash reference, a `types` or `typeRoots` inclusion in your tsconfig, or a `files`/`include` inclusion in your tsconfig.

If you’re proposing a syntax that can get around the only methods of file inclusion we have, that’s a bit of a paradox, isn’t it? Some file in your node_modules says that it should be included automatically without ever being referenced or imported—but how are we supposed to know it says that until it’s been referenced or imported? 🙃

**Comment by nbouvrette** [comment]
Hold on I think I a missing something @andrewbranch - what you are referring to works inside a package but my question was between packages. Let me try to rephrase this:

Package A (Text package including the `*.txt`) declaration is added as a dependency of Package B ("file API")

Package B does not need to import the file that includes the `*.txt` type declaration but needs to use `import text from './example.txt` and get the correct type.

I was thinking that maybe by installing the dependency, the global type could apply... (of course this is not the case today).

But thinking about it, in that case, if you want to use Package A as a Webpack loader actually to be able to import the files, if we refer to the declaration there it should actually work.

**Comment by andrewbranch** [comment]
What I said also works between packages, but the paradox still applies. Simply installing a dependency or having one in your node_modules is not generally sufficient for TypeScript to have any idea of what’s in there. But nothing is stopping you from adding `"package-a"` to your `types` array in tsconfig.json, and Package A providing that pattern ambient module declaration from its `main`/`types`/`exports` package.json entrypoint.

> by installing the dependency, the global type could apply

This is the problem. What you want can’t be achieved by adding new syntax. TypeScript is simply not a participant in the process of installing a dependency, and we very intentionally don’t trawl through 100s of MBs of node_modules looking for something that we might want to pull in without a configuration telling us to do so.
